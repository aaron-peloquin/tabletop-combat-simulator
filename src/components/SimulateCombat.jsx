import {connect} from 'react-redux'
import {Button} from '@material-ui/core'

import runSimulation from './../store/action/RunSimulation'

const SimulateCombat = (props) => {
  const {
    /** Actions */
    funcSimulateCombat,

    /** Props from State */
    Creatures,
  } = props
  return <Button
    data-cy="SimulateCombat"
    variant="contained"
    color="primary"
    onClick={()=>{
      funcSimulateCombat(Creatures)
    }}>Run Combat Simulation</Button>
}

const MapStateToProps = (state) => {
  const Creatures = state.creatures
  return {
    Creatures,
  }
}

const MapActionsToProps = (dispatch) => {
  return {
    funcSimulateCombat: (Creatures) => {
      runSimulation(dispatch, Creatures)
    },
  }
}

export default connect(MapStateToProps, MapActionsToProps)(SimulateCombat)
