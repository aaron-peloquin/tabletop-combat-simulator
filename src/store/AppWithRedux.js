import React from 'react'
import {initializeStore} from './store'

const isServer = typeof window === 'undefined'
const reduxStoreSymbol = Symbol("reduxStore")

function getOrCreateStore (initialState) {
  // Always make a new store if server, otherwise state is shared between requests
  if (isServer) {
    return initializeStore(initialState)
  }

  // Create store if unavailable on the client and set it on the window object
  if (!window[reduxStoreSymbol]) {
    window[reduxStoreSymbol] = initializeStore(initialState)
  }
  return window[reduxStoreSymbol]
}

export default (App) => {
  return class AppWithRedux extends React.Component {
    static async getInitialProps (appContext) {
      // Get or Create the store with `undefined` as initialState
      // This allows you to set a custom default initialState
      const reduxStore = getOrCreateStore()

      // Provide the store to getInitialProps of pages
      appContext.ctx.reduxStore = reduxStore

      let appProps = {}
      if (typeof App.getInitialProps === 'function') {
        appProps = await App.getInitialProps(appContext)
      }

      return {
        ...appProps,
        initialReduxState: reduxStore.getState()
      }
    }

    constructor (props) {
      super(props)
      this.reduxStore = getOrCreateStore(props.initialReduxState)
    }

    render () {
      return <App {...this.props} reduxStore={this.reduxStore} />
    }
  }
}
