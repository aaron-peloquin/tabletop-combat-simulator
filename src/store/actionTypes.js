/**
 * This is an object of Redux Actions. It mostly exists to
 *  ensure we do not create duplicate actions, but it can
 *  also be used as a list of our existing actions
*/
const actions = {
  CopyCreature: 'CopyCreature',
  DeleteAllCreatures: 'DeleteAllCreatures',
  DeleteCreature: 'DeleteCreature',
  SaveCreature: 'SaveCreature',
  SetEditCreature: 'SetEditCreature',
  ToggleSidebar: 'ToggleSidebar',
  RunSimulation: 'RunSimulation',
  UpdateEditCreature: 'UpdateEditCreature',
}

export default actions
