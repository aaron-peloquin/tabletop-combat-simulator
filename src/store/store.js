import {createStore, applyMiddleware, combineReducers} from "redux"
import {composeWithDevTools} from "redux-devtools-extension"
import thunkMiddleware from "redux-thunk"

/** Reducers */
import combat from "./reducer/combat"
import creatures from "./reducer/creatures"
import sideBar from "./reducer/sideBar"
import editing from "./reducer/editing"

/** Reducers */
const reducers = combineReducers({
  combat,
  creatures,
  editing,
  sideBar,
})

/**
 * Initializes the store
 * @param {obj} initialState the state our app starts with
 * @return {obj} the redux store
 */
export function initializeStore(initialState = initialState) {
  return createStore(reducers, initialState, composeWithDevTools(applyMiddleware(thunkMiddleware)))
}
