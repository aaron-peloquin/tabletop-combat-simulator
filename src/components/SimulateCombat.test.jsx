import {Provider} from "react-redux"
import {createShallow} from "@material-ui/core/test-utils"
import toJson from "enzyme-to-json"
import {initializeStore} from "../store/store"

import SimulateCombat from "../components/SimulateCombat"

describe("<SimulateCombat />", ()=>{
  const renderShallowUntilComponent = createShallow({"untilSelector": "SimulateCombat"})
  let store; let props; let SimulateCombatComponent
  beforeEach(()=>{
    store = initializeStore({})
    store.dispatch = jest.fn(store.dispatch)
    SimulateCombatComponent = renderShallowUntilComponent(<Provider store={store}><SimulateCombat {...props} /></Provider>)
  })

  it("calls dispatch on click", () => {
    SimulateCombatComponent.simulate("click")
    expect(store.dispatch).toBeCalledTimes(1)
  })

  it("snapshots", () => {
    const Tree = toJson(SimulateCombatComponent)
    expect(Tree).toMatchSnapshot()
  })
})
