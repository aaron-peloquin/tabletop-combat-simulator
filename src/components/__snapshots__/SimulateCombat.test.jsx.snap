// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SimulateCombat /> snapshots 1`] = `
<WithStyles(Button)
  color="primary"
  data-id="delete-all"
  onClick={[Function]}
  variant="contained"
>
  Run Combat Simulation
</WithStyles(Button)>
`;
